v4/service.go

```diff
@@ -0,0 +1,81 @@
+package suture
+
+import (
+	"context"
+	"errors"
+)
+
+/*
+Service is the interface that describes a service to a Supervisor.
+
+Serve Method
+
+The Serve method is called by a Supervisor to start the service.
+The service should execute within the goroutine that this is
+called in, that is, it should not spawn a "worker" goroutine.
+If this function either returns error or panics, the Supervisor
+will call it again, unless it returns special error sentinel values
+
+A Serve method SHOULD do as much cleanup of the state as possible,
+to prevent any corruption in the previous state from crashing the
+service again.
+
+The Supervisor will pass a ctx down to the service. That context
+can not be independently used to cancel the service, as the supervisor
```

Comment:
You can't actually cancel a context that was passed in to you, as you don't have a reference to the cancel function, so I am not sure this explanation even makes sense.

I think the important thing to note here is that the context is used for indication of the lifetime of the service, namely when the done channel is closed, the service should terminate

Reply:
I read this as “if you derive a subcontext using `WithCancel`, don’t try to cancel with the resulting `CancelFunc`”.

Reply:
Why not? That cancellation will not propagate upwards.

The whole point of passing a context down to the service is so that it could abort all ongoing http requests and what not as the service is asked to terminate, to ensure prompt termination, so you definately want to pass down the context deeper, create child contexts, and by definition you always have to cancel the child contexts in order not to have leaks (even if the task scoped by the context is done, in which cancellation is just resource cleanup).

Reply:
I believe Audrius was correct in the first comment. I was simply mistaken because the child service won't get the cancel function passed to it.

Thank you both for going over the docs; I was just expecting to do it myself. I will hopefully get to reviewing these changes late today my time (I'm in EST).